File: src/com/machfour/macros/data/Table.java
package com.machfour.macros.data;

import com.machfour.macros.core.MacrosPersistable;

import java.util.List;
import java.util.Map;

public interface Table<T extends MacrosPersistable> {
    String name();

    List<Column<?>> columns();

    Map<String, Column<?>> columnsByName();

    Column<?> columnForName(String name);

    T construct(ColumnData<T> dataMap, boolean isFromDb);
}

File: libmacros/src/main/kotlin/com/machfour/macros/sql/Table.kt
package com.machfour.macros.sql

import com.machfour.macros.core.EntityId
import com.machfour.macros.core.MacrosEntity
import com.machfour.macros.sql.entities.Factory

interface Table<I: MacrosEntity, M : I>: Factory<I, M> {
    val sqlName: String
    val columns: List<Column<M, out Any>>

    val columnsByName: Map<String, Column<M, *>>

    val idColumn: Column<M, EntityId>
    val createTimeColumn: Column<M, Long>
    val modifyTimeColumn: Column<M, Long>
}